File: go/day1/p2/p2.go
package p2

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"time"
)

const inputFile = "input.txt"  // Change to input.txt for final solution
const ignoreEmptyLines = false // Don't parse blank lines

func must[T any](v T, e error) T {
	if e != nil {
		panic(e)
	}
	return v
}

// #region Structs
type Input []int

// #endregion

// Parse
func parse() (elves Input) {
	// Open scanner to read input line by line
	scanner := bufio.NewScanner(must(os.Open(inputFile)))

	// Parsing state vars go here (if any)
	c := 0
	for scanner.Scan() {
		l := scanner.Text()
		if len(l) == 0 {
			elves = append(elves, c)
			c = 0
			continue
		}

		cals := must(strconv.Atoi(l))
		c += cals
	}
	return elves
}

// Solve
func solve(elves Input) (solution string) {
	// sum top 3
	sort.Slice(elves, func(i, j int) bool {
		return elves[i] < elves[j]
	})
	sum := 0
	for _, c := range elves[len(elves)-3:] {
		sum += c
	}
	return strconv.Itoa(sum)
}

func Part2() {
	// Parse
	elves := parse()

	// Solve
	start := time.Now()
	solution := solve(elves)

	// Report solve time and solution
	duration := time.Now().Sub(start)
	fmt.Printf("Solved in \x1b[34m%s\x1b[0m\n", duration)
	fmt.Printf("Solution: \x1b[32m%s\x1b[0m\n", solution)
}

File: rust/day1/src/p2.rs
use std::{io::{BufReader, BufRead}, fs::File, error::Error, time::Instant};

const INPUT_FILE: &str = "input.txt";

// #region Structs
type Input = Vec<i32>;

// #endregion

// Parse 
fn parse() -> Result<Input, Box<dyn Error>> {
  // Open scanner to read input line by line
  let file = File::open(INPUT_FILE)?;
  let reader = BufReader::new(file);
  
  // Parsing state vars go here (if any)
  let mut c = 0;
  let mut elves: Input = Vec::new();
  for l in reader.lines().map(|l| -> String {
    l.unwrap_or(String::new())
  }) {
    if l.len() == 0 {
      elves.push(c);
      c = 0;
      continue;
    }

    let cals: i32 = l.parse()?;
    c += cals;
  }
  Ok(elves)
}

// Solve
fn solve(input: &mut Input) -> String {
  // sum top 3 elves
  input.sort();
  let mut sum = 0;
  for c in input[input.len()-3..].iter() {
    sum += c;
  }
  format!("{}", sum)
}

pub fn part2() {
  // Parse
  let mut input = parse().expect("Failed to parse input");

  // Solve
  let start = Instant::now();
  let solution = solve(&mut input);

  // Report solve time and solution
  let duration = start.elapsed();
  println!("Solved in \x1b[34m{:?}\x1b[0m", duration);
  println!("Solution: \x1b[32m{}\x1b[0m", solution)
}